README.md
Lightberry Config Download
==========================

Simple plugin to easily download hyperion/boblight config files. 

What it has to offer:
---------------------
   * Downloads hyperion/boblight configs for different led placements (horizontal x vertical)
   * Differentiate between OpenELEC (different config and effects placement) and other systems
   * Replace grabber section (causes hyperion to fail if there is no /dev/video device)

How to install:
---------------
   1. Download latest release from [github releases](https://github.com/tszczerba/lightberry-config-downloader/releases)
   2. Copy to raspberry pi or make avaliable in local network (you can use smb e.g. \\raspbmc or \\openelec)
   3. From XBMC menu go **System** -> **Add-ons** -> **Install from zip file** -> *Find zip file from previous point and install it*
   4. Plugin should appear under **Programs** menu
   5. For easy access go **System** -> **Appearance** -> **Skin** -> **Settings** *Click on what appears to be just a spacer* -> **Add-on Shortcuts** -> **Home Page Programs Sub-menu** -> *Set plugin on desired position* (first two are already taken by raspbmc specific add-ons if you are running raspbmc, so pick 3rd or forward place in this case)

Please create issues, functionality requests here - I can use any sugestion, I'm python begginner. Thank you in advance.


changelog.txt
1.2.0
- Add copy from USB Pendrive for model A

1.1.0
- Add grabber section to OpenELEC download

1.0.7
- Fixed OpenELEC catalog

1.0.6
- Add replace grabber-v4l2 section

1.0.5
- Download logic change

1.0.4
- More fixes for OpenELEC case

1.0.3
- Fixes for OpenELEC case

1.0.2
- Support for OpenELEC

1.0.1
- New folder structure for configs

1.0.0
- initial version

addon.py
#  Copyright (C) 2014 lightberry.eu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with XBMC; see the file COPYING.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
import sys
import os

import xbmcaddon
import xbmc
from resources.lib import ConfigDownload, HyperionControl


__author__ = "Tomek (lightberry.eu)"
__url__ = "http://lightberry.eu"
__addonname__ = "Lightberry Config Downloader"
__svn_url__ = ""
__credits__ = "Hyperion Switcher plugin for Raspbmc written and maintained by lightberry.eu"
__icon__ = "./icon.png"

__settings__ = xbmcaddon.Addon()
__addondir__ = __settings__.getAddonInfo('path')

delayTime = 5000
msgLine = ""
exceptionLine = ""
devVideoName = "/dev/video0"

ledConfigPrev = __settings__.getSetting("ledConfig")
ledControlSystemPrev = __settings__.getSetting("ledControlSystem")

if ConfigDownload.addonConfigUpdate(__addondir__):
    msgLine = "Successfully downloaded new addon configuration."
    xbmc.executebuiltin('Notification(%s, %s, %d, %s)' % (__addonname__, msgLine, delayTime, __icon__))
else:
    msgLine = "Addon Configuration is up to date."
    xbmc.executebuiltin('Notification(%s, %s, %d, %s)' % (__addonname__, msgLine, delayTime, __icon__))

try:
    __settings__.openSettings()
except KeyboardInterrupt, SystemExit:
    sys.exit(0)

if __settings__.getSetting("fileSource") == "Internet":
    if __settings__.getSetting("downloadNow") == "true" \
            or ledConfigPrev != __settings__.getSetting("ledConfig")\
            or ledControlSystemPrev != __settings__.getSetting("ledControlSystem"):
        try:
            ConfigDownload.downloadConfig(__settings__.getSetting("ledConfig"),
                                          __settings__.getSetting("ledControlSystem"))

            if os.uname()[1] != "OpenELEC" :
                msgLine = "Successfully downloaded configuration " + __settings__.getSetting(
                "ledConfig") + " for " + __settings__.getSetting("ledControlSystem")

            else:
                msgLine = "Successfully downloaded configuration " + __settings__.getSetting(
                "ledConfig") + " for " + __settings__.getSetting("ledControlSystem") + ".\n" \

        except:
            msgLine = "Error occurred. Check your internet connection."

        xbmc.executebuiltin('Notification(%s, %s, %d, %s)' % (__addonname__, msgLine, delayTime, __icon__))

        if ((not os.path.exists(devVideoName) and __settings__.getSetting("grabberReplaceLogic") == "Default") \
                or __settings__.getSetting("grabberReplaceLogic") == "Force replace") \
                and __settings__.getSetting("ledControlSystem") == "hyperion":

            ConfigDownload.replaceGrabberSection()

            msgLine = "Grabber section replaced "
            xbmc.executebuiltin('Notification(%s, %s, %d, %s)' % (__addonname__, msgLine, delayTime, __icon__))
else:
    try:
        ConfigDownload.copyFromUsb(__settings__.getSetting("ledControlSystem"))
        msgLine = "Configuration File successfully copied."
    except:
        msgLine = "No configuration file found."

    xbmc.executebuiltin('Notification(%s, %s, %d, %s)' % (__addonname__, msgLine, delayTime, __icon__))

if os.uname()[1] != "OpenELEC":
    hss = HyperionControl.HyperionControl()
    hss.service("restart")

__settings__.setSetting("downloadNow","false")

resources/lib/HyperionControl.py
#  Copyright (C) 2014 lightberry.eu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with XBMC; see the file COPYING.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html

import subprocess
import os

__author__ = 'lightberry.eu'


class HyperionControl:
    currentHyperionSource = ""

    hyperiondName = "hyperiond"
    hyperionPid = None

    hyperiondServiceCommand = "service hyperion"

    def __init__(self):
        self.processNum()

    def processNum(self):
        if os.uname()[1] == "raspbmc":
            command = 'ps -eo pid,command | grep "{0:s}" | grep -v grep | awk \'{{print $1}}\''.format(self.hyperiondName)
        elif os.uname()[1] == "OpenELEC":
            command = 'ps -e pid,command | grep "{0:s}" | grep -v grep | awk \'{{print $1}}\''.format(self.hyperiondName)
        else :
            command = 'ps -eo pid,command | grep "{0:s}" | grep -v grep | awk \'{{print $1}}\''.format(self.hyperiondName)
        self.hyperionPid = subprocess.check_output(
             command,
            shell=True)

    def service(self, mode):
        if os.uname()[1] == "OpenELEC":
            sudoRequired = False
        else :
            sudoRequired = True

        try:
            if mode.lower() == "start" or mode.lower() == "restart":
                self.processNum()
                if self.hyperionPid == "":
                    self.execute(self.hyperiondServiceCommand + " start", sudo=sudoRequired)
                else:
                    self.execute(self.hyperiondServiceCommand + " restart", sudo=sudoRequired)
            elif mode.lower == "stop":
                self.processNum()
                if self.hyperionPid != "":
                    self.execute(self.hyperiondServiceCommand + " stop", sudo=sudoRequired)
        except subprocess.CalledProcessError:
            raise Exception("Something went wrong with hyperion service. \nIs hyperion installed?")

    @staticmethod
    def execute(command, sudo=False):
        if sudo:
            subprocess.check_call("sudo " + command, shell=True)
        else:
            subprocess.check_call(command, shell=True)

    pass

resources/lib/ConfigDownload.py
#  Copyright (C) 2014 lightberry.eu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with XBMC; see the file COPYING.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
import urllib
import os
import re
import subprocess

infoStr = "/// ***\n" \
          "        /// Section for \"grabber-v4l2\" commented out.\n" \
          "        /// Configure v4l to use this feature\n"

tempFolder = "/tmp/"

configOptions = {'hyperion': 'hyperion.config.json', 'boblight': 'boblight.conf'}


def downloadConfig(config, configFor):
    if os.uname()[1] == "raspbmc":
        lightberryRepoAddress = "http://lightberry.eu/download/General/"
        configFolder = "/etc/"
        replaceEffectsInd = False
    elif os.uname()[1] == "OpenELEC":
        lightberryRepoAddress = "http://lightberry.eu/download/General/"
        configFolder = "/storage/.config/"
        replaceEffectsInd = True
    else :
        lightberryRepoAddress = "http://lightberry.eu/download/General/"
        configFolder = "/etc/"
        replaceEffectsInd = False

    fileAddress = lightberryRepoAddress + config + "/" + configOptions[configFor]

    tempFile = tempFolder + configOptions[configFor]
    destFile = configFolder + configOptions[configFor]
    bkpFile = destFile + "_bkp"

    urllib.urlretrieve(fileAddress, tempFile)

    if replaceEffectsInd:
        replaceEffectsFolder(tempFile)

    mvFromTo(destFile,bkpFile)
    mvFromTo(tempFile,destFile)

def replaceEffectsFolder(tfile):
    tempFile = tfile + "_replace"
    destFile = tfile

    with open(destFile, 'r') as f:
        with open(tempFile, 'w+') as t:
            t.write(f.read().replace('"/opt/hyperion/effects"','"/storage/hyperion/effects"',1))

    mvFromTo(tempFile,destFile)

def execute(command):
    if os.uname()[1] == "raspbmc":
        s = True
    elif os.uname()[1] == "OpenELEC":
        s = False
    else :
        s = False

    if s:
        child = subprocess.Popen("sudo " + command, shell=True)
    else:
        child = subprocess.Popen(command, shell=True)
    child.wait()

def mvFromTo(f,t):
    execute("mv " + f + " " + t)
    execute("chmod 755 " + t)
    execute("chown root:root " + t)

def cpFromTo(f,t):
    execute("cp " + f + " " + t)
    execute("chmod 755 " + t)
    execute("chown root:root " + t)

def addonConfigUpdate(addonDir):
    lightberryAddonConfigLocalVersionFile = addonDir + "/resources/version.info"
    lightberryAddonConfigLocal = addonDir + "/resources/settings.xml"

    lightberryAddonConfigRemoteVersionFile = "http://img.lightberry.eu/xbmcAddons/config-downloaderV2/version.info"
    lightberryAddonConfigRemoteVersionTmp = addonDir + "/resources/version.info.remote"
    lightberryAddonConfigRemote = "http://img.lightberry.eu/xbmcAddons/config-downloaderV2/settings.xml"

    urllib.urlretrieve(lightberryAddonConfigRemoteVersionFile, lightberryAddonConfigRemoteVersionTmp)

    with open(lightberryAddonConfigRemoteVersionTmp, 'r') as f:
        remoteVersion = f.readline()

    with open(lightberryAddonConfigLocalVersionFile, 'r') as f:
        localVersion = f.readline()

    if remoteVersion != localVersion:
        urllib.urlretrieve(lightberryAddonConfigRemote, lightberryAddonConfigLocal)
        urllib.urlretrieve(lightberryAddonConfigRemoteVersionFile, lightberryAddonConfigLocalVersionFile)
        return True
    else:
        return False

def replaceGrabberSection():
    if os.uname()[1] == "raspbmc":
        configFolder = "/etc/"
    elif os.uname()[1] == "OpenELEC":
        configFolder = "/storage/.config/"
    else :
        configFolder = "/etc/"

    tempFile = tempFolder + configOptions["hyperion"]
    destFile = configFolder + configOptions["hyperion"]

    with open(destFile, 'r') as f:
        with open(tempFile, 'w+') as t:
            t.write(re.sub(r"(\"grabber-v4l2\")+(.*)\n((.*)\{([\s\S]*?)\}(.*),)",infoStr, f.read(), re.MULTILINE))

    mvFromTo(tempFile, destFile)

def copyFromUsb(configFor):
    if os.uname()[1] == "raspbmc":
        configFolder = "/etc/"
    elif os.uname()[1] == "OpenELEC":
        configFolder = "/storage/.config/"
    else :
        configFolder = "/etc/"

    destFile = configFolder + configOptions[configFor]

    filePath = findFileToCopy(configOptions[configFor])
    if filePath != "":
        cpFromTo(filePath,destFile)
    else:
        raise Exception("File not found")

def find(name, path):
    for root, dirs, files in os.walk(path):
        if name in files:
            return os.path.join(root, name)

    return ""

def findFileToCopy(fileName):
    try:
        usbMediaPath = "/media"
        return find(fileName, usbMediaPath)
    except:
        return ""